v3/bind.go

```diff
@@ -733,3 +733,115 @@ RESP:
 
 	return nil, GetLDAPError(packet)
 }
+
+// ServerBindStep sends the SASLBindRequest and return the result code, servercred and errors.
+// If the result code is not SUCCESS(0x00) or IN_PROGRESS(0x0e), it's also considered as an error.
+func (l *Conn) ServerBindStep(clientCred []byte, dn string, methodName string, controls []Control) (resultCode uint16, serverCred []byte, err error) {
+	packet := ber.Encode(ber.ClassUniversal, ber.TypeConstructed, ber.TagSequence, nil, "LDAP Request")
+	packet.AppendChild(ber.NewInteger(ber.ClassUniversal, ber.TypePrimitive, ber.TagInteger, l.nextMessageID(), "MessageID"))
+
+	request := ber.Encode(ber.ClassApplication, ber.TypeConstructed, ApplicationBindRequest, nil, "Bind Request")
+	request.AppendChild(ber.NewInteger(ber.ClassUniversal, ber.TypePrimitive, ber.TagInteger, 3, "Version"))
+	request.AppendChild(ber.NewString(ber.ClassUniversal, ber.TypePrimitive, ber.TagOctetString, dn, "User Name"))
+
+	auth := ber.Encode(ber.ClassContext, ber.TypeConstructed, 3, "", "authentication")
+	auth.AppendChild(ber.NewString(ber.ClassUniversal, ber.TypePrimitive, ber.TagOctetString, methodName, "SASL Mech"))
+	auth.AppendChild(ber.NewString(ber.ClassUniversal, ber.TypePrimitive, ber.TagOctetString, string(clientCred), "Credentials"))
+	request.AppendChild(auth)
+	packet.AppendChild(request)
+
+	if len(controls) > 0 {
+		packet.AppendChild(encodeControls(controls))
+	}
+
+	msgCtx, err := l.sendMessage(packet)
+	if err != nil {
+		return 0, nil, err
+	}
+	defer l.finishMessage(msgCtx)
+
+	packetResponse, ok := <-msgCtx.responses
+	if !ok {
+		return 0, nil, errors.New("ldap: response channel closed")
+	}
+	packet, err = packetResponse.ReadPacket()
+	if err != nil {
+		return 0, nil, err
+	}
+
+	bindResponse, err := parseBindResponse(packet)
+	if err != nil {
+		return 0, nil, err
+	}
+
+	// TODO: add global constants. As this package used `uint16` to represent the result code nearly everywhere,
+	// these constants were
+	const LDAPResultCodeSuccess = 0
+	const LDAPResultCodeSASLBindInProgress = 0x0e
```

Comment:
Both values are already defined in `error.go`, thought they're of type int unfortunately. I think it can stay like this and we take care of this later on 😄